README.rst

```diff
@@ -1,103 +1,262 @@
+   - ``f``: **float**
+   - ``s``:**string**
+   - ``*``: complex (see below)
+- 1 optional letter specifying endianness:
+   - *default*: big endian
+   - ``b``: **big** endian
+   - ``l``: **little** endian
+- 1 optional letter specifying big-endianness for unaligned reads:
+   - *default*: left -> right
+   - ``n``: left -> right (**normal**)
+   - ``r``: left <- right (**reverse**)
+- 1 number specifying size in **bits**:
+   - for strings only byte multiples are allows (``8``, ``16``, ``24``, ...)
+   - for an integer the allowed values are ``1 .. 64``
+   - for a float the allowed values are ``32`` and ``64``
+   - for a custom it can't be used (use the secondary ``size`` operation)
+
+| Strings may not have a size, in which case they are null-terminated.
+| You can order options however you want, but size must come last (e.g. ``lru16`` and ``url16`` are valid but not ``16lru``).
+
+Value
+~~~~~
+
+This section consists of a mandatory name for the field + optional
+repetition + optional assertion.
+
+For primitive types (except null-terminated strings) you can instruct
+binaryparse to not produce a symbol and discard the field by using ``_``
+for the name:
+
+.. code:: nim
+
+    createParser(magic):
+      s: _ = "GIF89a"
+
+Alignment
+~~~~~~~~~
+
+Currently unaligned reads for strings are not supported:
+
+.. code:: nim
+
+    createParser(myParser):
+      6: x
+      s: y # invalid, generates an exception
+
+| Endian only refers to aligned reads, while bit-endian only refers to unaligned reads.
+| When reading unaligned data, you should always ensure that bit-endian is uniform between **byte boundaries**.
+| Your spec must finish on a byte boundary.
+
+.. code:: nim
+
+    createParser(myParser, bitEndian = n):
+      2: a
+      16: b # n/r will be considered; l/b is irrelevant
+      r6: c # undefined behavior: shares bits with previous byte
+      16: d # l/b will be considered; n/r is irrelevant
+      10: e # undefined behavior: spec does not finish on a byte boundary
+
+-  When using an irrelevant option, binaryparse should generate a
+   **warning**
+-  When switching bit-endian between 2 unaligned reads binaryparse
+   should generate an **exception**
+-  When spec does not end on a byte boundary, binaryparse should
+   either generate an **exception** or pad the last byte with
+   zeros and generate a **warning**
+
+Repetition
+~~~~~~~~~~
+
+There are 2 ways to produce a ``seq`` of your ``Type``:
+
+- ``for repetition``: append ``[expr]`` to the name for repeating ``expr``
+  times
+- ``until repetition``: append ``{expr}`` to the name for repeating until
+  ``expr`` is evaluated to ``true``
+
+In until repetition you can use 3 special symbols:
+
+- ``e``: means 'last element read'
+- ``i``: means 'current loop index'
+- ``s``: means 'stream'
+
+.. code:: nim
+
+    u8: a{e == 103 or i > 9} # reads until it finds the value 103 or completes 10th iteration
+
+.. code:: nim
+
+    16: x[5] # seq[int16] of size 5
```

Comment:
This should also show how `[]` can be used with previous fields and without an argument for "loop until magic sequence".

Reply:
This is changed. Checkout "assertion". We no longer need this. The same can be done with:
```nim
16: x{s.atEnd}
```

Reply:
for "loop until magic sequence" I propose the syntax `{symbol}` instead of `symbol[]`. I think it is more intuitive and also it's easier to parse. For example you can even do `8: {symbol[expr]}` to get a `seq[seq[int8]]`.